modyn/common/trigger_storage_cpp/trigger_storage_cpp.py

```diff
@@ -0,0 +1,433 @@
+        size = (ctypes.c_uint64 * 1)()
+        self.data_pointer.shape_val = size
+        pattern = ctypes.c_char_p(f"{pipeline_id}_{trigger_id}_{partition_id}_".encode("utf-8"))
+
+        data = self._get_worker_samples_impl(folder, size, pattern, start_index, worker_subset_size).reshape(-1)
+        result = ArrayWrapper(data, self._release_data_impl)
+        return result
+
+    def _get_all_samples(self, pipeline_id: int, trigger_id: int, partition_id: int) -> ArrayWrapper:
+        """
+        Return all the samples for the given pipeline id, trigger id and partition id.
+
+        :param pipeline_id: the id of the pipeline
+        :param trigger_id: the id of the trigger
+        :param partition_id: the id of the partition
+        :return: the trigger samples
+        """
+
+        folder = ctypes.c_char_p(str(self.trigger_sample_directory).encode("utf-8"))
+        size = (ctypes.c_uint64 * 1)()
+        self.data_pointer.shape_val = size
+        pattern = ctypes.c_char_p(f"{pipeline_id}_{trigger_id}_{partition_id}_".encode("utf-8"))
+
+        data = self._get_all_samples_impl(folder, size, pattern).reshape(-1)
+        result = ArrayWrapper(data, self._release_data_impl)
+        return result
+
+    def save_trigger_sample(
+        self, pipeline_id: int, trigger_id: int, partition_id: int, trigger_samples: np.ndarray, insertion_id: int
+    ) -> None:
+        """
+        Save the trigger samples for the given pipeline id, trigger id and partition id.
+
+        :param pipeline_id: the id of the pipeline
+        :param trigger_id: the id of the trigger
+        :param partition_id: the id of the partition
+        :param trigger_samples: the trigger samples
+        :param insertion_id: the id of the insertion
+        """
+        if trigger_samples.dtype != np.dtype("i8,f8"):
+            raise ValueError(f"Unexpected dtype: {trigger_samples.dtype}\nExpected: {np.dtype('i8,f8')}")
+
+        Path(self.trigger_sample_directory).mkdir(parents=True, exist_ok=True)
+
+        samples_file = Path(self.trigger_sample_directory) / f"{pipeline_id}_{trigger_id}_{partition_id}_{insertion_id}"
+
+        assert not Path(samples_file).exists(), (
+            f"Trigger samples file {samples_file} already exists. " f"Please delete it if you want to overwrite it."
+        )
+
+        self._write_file(samples_file, trigger_samples)
+
+    def save_trigger_samples(
+        self, pipeline_id: int, trigger_id: int, partition_id: int, trigger_samples: np.ndarray, data_lengths: list
+    ) -> None:
+        """
+        Save the trigger samples for the given pipeline id, trigger id and partition id
+        to multiple files.
+
+        :param pipeline_id: the id of the pipeline
+        :param trigger_id: the id of the trigger
+        :param partition_id: the id of the partition
+        :param trigger_samples: the trigger samples
+        :param data_lengths: the lengths of subarrays to write to files
```

Comment:
Right now it's as swap-in-swap-out as possible with the previous version, but I'll move more of the threading loop into the trigger storage function so that there is less code duplication.